src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,1209 @@
+                    await Task.Delay(1000);
+                    if(sp.IsRunning)
+                        sp.Reset();
+
+                    continue;
+                }
+
+                var rejectedShards = _rejectedShards;
+                if (CancellationTokenSource.IsCancellationRequested)
+                    break;
+
+                if (sp.IsRunning == false)
+                    sp.Start();
+      
+                if (sp.Elapsed >= _rejectedReconnectTimeout)
+                    throw new SubscriptionConnectionDownException($"Could not reconnect shards '{string.Join(",", rejectedShards)}' stopping sharded subscription '{Options.SubscriptionName}'.");
+
+                var workers = new List<SubscriptionServerWorker>();
+                var skippedNodes = new Dictionary<string, List<string>>();
+                foreach (var shard in rejectedShards)
+                {
+                    skippedNodes[shard] = new List<string>();
+                    var shardId = ShardHelper.TryGetShardIndex(shard);
+                    var re = TcpConnection.ShardedContext.RequestExecutors[shardId];
+                    if (re.Topology == null)
+                    {
+                        await re.WaitForTopologyUpdate(re._firstTopologyUpdate);
+                        if (re.Topology == null)
+                        {
+                            throw new Exception("ShouldNot happen and likely a bug.!!");
+                        }
+                    }
+
+                    SubscriptionServerWorker worker = null;
+                    List<Exception> exceptions = new List<Exception>();
+                    foreach (var node in re.Topology.Nodes)
+                    {
+                        try
+                        {
+                            worker = await GetShardedSubscriptionServerWorker(shard, re, node.ClusterTag);
+                        }
+                        catch (Exception ex)
+                        {
+                            exceptions.Add(ex);
+                            skippedNodes[shard].Add(node.ClusterTag);
+                            continue;
+                        }
+
+                        workers.Add(worker);
+                        break;
+                    }
+                }
+
+                if (workers.Count == 0)
+                    continue;
+
+                foreach (var worker in workers)
+                {
+                    await SendOptionToServerWorker(worker);
+                }
+
+                if (CancellationTokenSource.IsCancellationRequested)
+                    break;
+
+                await AssertConnectionStatusForRejectedConnection(workers, skippedNodes, sp);
+            }
+        }
+
+        //TODO: egor handle double code with InitAsync and AssertConnection
+        private async Task AssertConnectionStatusForRejectedConnection(ICollection<SubscriptionServerWorker> workers, IReadOnlyDictionary<string, List<string>> skippedNodes, Stopwatch stopwatch)
```

Comment:
All of this should just be the usual connection trying to connect to the shard and running all the usual code.
Why duplicate it here?